Add "base dicts"

cassella opened the pull request:
One or more --base-dict's may be specified on the command line.  We
search for tokens in these dicts, but do not add words to them.

--use-builtin-base-dict uses the dictionary.txt in SCSPELL_DATA_DIR as
a base dict.

This allows a project-specific dictionary to be maintained containing
only words from the project, without including the full 844K base
dictionary and its license terms.  E.g.

$ cd PROJECT
$ scspell --override-dictionary .scspell/dictionary.txt \
          --use-builtin-base-dict \
          $PROJECT_FILES

Another benefit is that as scspell's dictionary is updated over time,
scspell invoked with --use-builtin-base-dict will reflect the updated
dictionary automatically.

cassella committed:
Support for base dicts

One or more --base-dict's may be specified on the command line.  We
search for tokens in these dicts, but do not add words to them.

--use-builtin-base-dict uses the dictionary.txt in SCSPELL_DATA_DIR as
a base dict.

This allows a project-specific dictionary to be maintained containing
only words from the project, without including the full 844K base
dictionary and its license terms.  E.g.

$ cd PROJECT
$ scspell --override-dictionary .scspell/dictionary.txt \
          --use-builtin-base-dict \
          $PROJECT_FILES

Another benefit is that as scspell's dictionary is updated over time,
scspell invoked with --use-builtin-base-dict will reflect the updated
dictionary automatically.

Implementation notes:

The CorporaFile object for the main dictionary of a run is built with
a list of additional CorporaFile objects, one per base dict.

The list is in the order the base dicts were specified on the command
line.  The order doesn't matter -- the return value of match() depends
only on whether the word is in any of the dictionaries or in none of
them.

cassella committed:
Add --filter-out-base-dicts option

The intended use is for people who've built a custom dictionary using
an old version of scspell.  The custom dictionary will include the
entire wordlist we ship.  Invoking

scspell --use-builtin-base-dict --filter-base-dicts

will remove from the custom dict all the words that are in our shipped
list.  (With the intention that future invocations of scspell using
the custom dict will use --use-builtin-base-dict)

Implementation Note:

As described in the CorporaFile.filter_out_base_dicts() comment, this
function wants to match words to base dicts, but have control over
which worldlist(s) to check.  (natural vs. filetype vs. per-file.)
CorporaFile.match() now has a match_in parameter (defaulting to all)
to provide this control.

cassella committed:
README: Update for base dicts

cassella committed:
_util.detect_encoding(): If file doesn't exist, return "utf-8"

This is needed for open_with_encoding("newfile", "w", ...), since
open_with_encoding()->detect_encoding() opens the file for reading to
detect its encoding.

detect_encoding() returns "utf-8" for an empty file, so that seems
like the right choice for a file that doesn't exist yet also.

cassella committed:
Add --base-dicts tests

cassella commented:
This one should be more straightforward.  :)

Once this is merged, I thought I might change the scspell invocation in "make check" to use this feature, so that things like "pelzl" and "myint" could be added to a local .scspell/dictionary.txt.   That way every new developer can just run "make check" without building up their own ~/.scspell/dictionary.txt first.

BTW, why doesn't make check run pyflakes?

myint commented on scspell/_util.py:
```diff
@@ -60,8 +60,20 @@ def open_with_encoding(filename, encoding=None, mode='r'):
+
+    # Thanks again stackoverflow for python 2 v 3,
+    # http://stackoverflow.com/questions/15032108/pythons-open-throws-different-errors-for-file-not-found-how-to-handle-b
+    fnotfound = getattr(__builtins__, 'FileNotFoundError', IOError)
```
I think simply using `except (IOError, OSError):` would be much more elegant and easier to understand.

cassella replied:
My intention was for this to handle only file-not-found, and continue to propagate e.g. permission errors up the stack as before.

Though now that I think I understand that getattr() line a little better, I see that's not what it's doing.

myint commented:
I think we ought to make it run `pyflakes`.

myint commented on scspell/_corpus.py:
```diff
@@ -315,6 +334,47 @@ def match(self, token, filename, file_id):
+                newtokens.append(t)
+        self._natural_dict._tokens = newtokens
+
+        for ext in self._extensions.keys():
```
I think `for ext in self._extensions:` would be simpler.

cassella committed:
filter_out_base_dicts(): iterate over extensions, not extensions.keys()

cassella committed:
detect_encoding: simplify error handling.

If it's some error other than ENOENT, we'll report it for real when
we actually go to open the file in order to use it.

myint commented:
Thanks.
